<?xml version="1.0"?>
<launch>

    <!-- Launch the node for communication with PCB -->
    <include file="$(find noah_drivers)/launch/drivers_node.launch">
        <arg name="uart_port_name" value="$(env PCB_UART_PORT)" />
        <arg name="uart_update_rate" value="$(env PCB_UART_UPDATE_RATE_HZ)"/>
        <arg name="pid_min" value="$(env PID_MIN)" />
        <arg name="pid_max" value="$(env PID_MAX)" />
    </include>

     <!-- Launch the nodes for PID control of the motors -->
    <rosparam subst_value="true">
        ticks_meter: $(env TICK_METERS)
    </rosparam>
    <node pkg="differential_drive" type="pid_velocity.py" name="motor_left_pid_velocity">
        <remap from="wheel" to="/uart_comms_node/motor_left_tick"/>
        <remap from="motor_cmd" to="/uart_comms_node/motor_left_pwm"/>
        <remap from="wheel_vtarget" to="rwheel_vtarget"/>
        <remap from="wheel_vel" to="motor_left_current_speed"/>
        <rosparam subst_value="true">
            Kp: $(env PID_P)
            Ki: $(env PID_I)
            Kd: $(env PID_D)
            out_min: -$(env PID_MAX)
            out_max: $(env PID_MAX)
            rate: $(env PCB_UART_UPDATE_RATE_HZ)
            timeout_ticks: 10
            rolling_pts: 5
        </rosparam>
    </node>

    <node pkg="differential_drive" type="pid_velocity.py" name="motor_right_pid_velocity">
        <remap from="wheel" to="/uart_comms_node/motor_right_tick"/>
        <remap from="motor_cmd" to="/uart_comms_node/motor_right_pwm"/>
        <remap from="wheel_vtarget" to="lwheel_vtarget"/>
        <remap from="wheel_vel" to="motor_right_current_speed"/>
        <rosparam subst_value="true">
            Kp: $(env PID_P)
            Ki: $(env PID_I)
            Kd: $(env PID_D)
            out_min: -$(env PID_MAX)
            out_max: $(env PID_MAX)
            rate: $(env PCB_UART_UPDATE_RATE_HZ)
            timeout_ticks: 10
            rolling_pts: 5
            encoder_min: $(env ENCODER_MIN)
            encoder_max: $(env ENCODER_MAX)
        </rosparam>
    </node>
  
    <node pkg="differential_drive" type="twist_to_motors.py" name="twist_to_motors" output="screen">
        <rosparam subst_value="true" param="base_width">$(env BASE_WIDTH)</rosparam>
        <rosparam subst_value="true" param="rate">$(env PCB_UART_UPDATE_RATE_HZ)</rosparam>
    </node>

    <node pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py" name="teleop_keyboard" output="screen">
        <param name="scale_linear" value="0.3"/>
        <param name="scale_angular" value="0.3"/>
        <remap from="cmd_vel" to="twist"/>
    </node>
</launch>